File: programs/svs-5/src/math.rs
//! Math module wrapper - re-exports from svs-math with Anchor error conversion.

use anchor_lang::prelude::*;

use crate::error::VaultError;
use crate::state::StreamVault;

// Re-export types from svs-math
pub use svs_math::Rounding;

/// Convert assets to shares with virtual offset protection against inflation attacks.
///
/// Wraps svs_math::convert_to_shares with Anchor error conversion.
pub fn convert_to_shares(
    assets: u64,
    total_assets: u64,
    total_shares: u64,
    decimals_offset: u8,
    rounding: Rounding,
) -> Result<u64> {
    svs_math::convert_to_shares(
        assets,
        total_assets,
        total_shares,
        decimals_offset,
        rounding,
    )
    .map_err(|e| match e {
        svs_math::MathError::Overflow => VaultError::MathOverflow.into(),
        svs_math::MathError::DivisionByZero => VaultError::DivisionByZero.into(),
    })
}

/// Convert shares to assets with virtual offset protection.
///
/// Wraps svs_math::convert_to_assets with Anchor error conversion.
pub fn convert_to_assets(
    shares: u64,
    total_assets: u64,
    total_shares: u64,
    decimals_offset: u8,
    rounding: Rounding,
) -> Result<u64> {
    svs_math::convert_to_assets(
        shares,
        total_assets,
        total_shares,
        decimals_offset,
        rounding,
    )
    .map_err(|e| match e {
        svs_math::MathError::Overflow => VaultError::MathOverflow.into(),
        svs_math::MathError::DivisionByZero => VaultError::DivisionByZero.into(),
    })
}

/// Safe multiplication then division with configurable rounding.
///
/// Wraps svs_math::mul_div with Anchor error conversion.
pub fn mul_div(value: u64, numerator: u64, denominator: u64, rounding: Rounding) -> Result<u64> {
    svs_math::mul_div(value, numerator, denominator, rounding).map_err(|e| match e {
        svs_math::MathError::Overflow => VaultError::MathOverflow.into(),
        svs_math::MathError::DivisionByZero => VaultError::DivisionByZero.into(),
    })
}

/// Calculate effective total assets including accrued streaming yield.
///
/// Returns the total assets available for share conversion, including
/// any yield that has accrued since the last checkpoint.
pub fn effective_total_assets(vault: &StreamVault, now: i64) -> Result<u64> {
    // If stream has ended or never started (stream_start >= stream_end), return full amount
    if now >= vault.stream_end || vault.stream_start >= vault.stream_end {
        return vault
            .base_assets
            .checked_add(vault.stream_amount)
            .ok_or(VaultError::MathOverflow.into());
    }
    
    // If before stream start, return base assets only
    if now <= vault.stream_start {
        return Ok(vault.base_assets);
    }
    
    // Calculate accrued yield
    let elapsed = (now - vault.stream_start) as u64;
    let duration = (vault.stream_end - vault.stream_start) as u64;
    let accrued = mul_div(vault.stream_amount, elapsed, duration, Rounding::Floor)?;
    
    vault
        .base_assets
        .checked_add(accrued)
        .ok_or(VaultError::MathOverflow.into())
}

File: sdk/svs-6/src/math.ts
import BN from "bn.js";

const ZERO = new BN(0);
const ONE = new BN(1);

/** Calculate the virtual offset for inflation protection. */
export function calculateOffset(decimalsOffset: number): BN {
  return new BN(10).pow(new BN(decimalsOffset));
}

/** Calculate accrued streaming yield at a given timestamp. */
export function calculateAccrued(
  streamAmount: BN,
  streamStart: BN,
  streamEnd: BN,
  currentTimestamp: BN
): BN {
  if (streamAmount.isZero() || streamEnd.lte(streamStart)) {
    return ZERO;
  }

  const duration = streamEnd.sub(streamStart);
  const elapsed = BN.max(currentTimestamp.sub(streamStart), ZERO);
  const cappedElapsed = BN.min(elapsed, duration);

  // accrued = streamAmount * cappedElapsed / duration
  return streamAmount.mul(cappedElapsed).div(duration);
}

/** Calculate effective total assets including streaming yield. */
export function effectiveTotalAssets(
  baseAssets: BN,
  streamAmount: BN,
  streamStart: BN,
  streamEnd: BN,
  currentTimestamp: BN
): BN {
  const accrued = calculateAccrued(
    streamAmount,
    streamStart,
    streamEnd,
    currentTimestamp
  );
  return baseAssets.add(accrued);
}

/** Convert assets to shares with virtual offset (floor rounding). */
export function convertToShares(
  assets: BN,
  totalAssets: BN,
  totalShares: BN,
  decimalsOffset: number
): BN {
  const offset = calculateOffset(decimalsOffset);
  const virtualShares = totalShares.add(offset);
  const virtualAssets = totalAssets.add(ONE);

  // shares = assets * virtualShares / virtualAssets (floor)
  return assets.mul(virtualShares).div(virtualAssets);
}

/** Convert shares to assets with virtual offset (floor rounding). */
export function convertToAssets(
  shares: BN,
  totalAssets: BN,
  totalShares: BN,
  decimalsOffset: number
): BN {
  const offset = calculateOffset(decimalsOffset);
  const virtualShares = totalShares.add(offset);
  const virtualAssets = totalAssets.add(ONE);

  // assets = shares * virtualAssets / virtualShares (floor)
  return shares.mul(virtualAssets).div(virtualShares);
}

/** Convert assets to shares with ceiling rounding (for withdraw). */
export function convertToSharesCeil(
  assets: BN,
  totalAssets: BN,
  totalShares: BN,
  decimalsOffset: number
): BN {
  const offset = calculateOffset(decimalsOffset);
  const virtualShares = totalShares.add(offset);
  const virtualAssets = totalAssets.add(ONE);

  // ceil(assets * virtualShares / virtualAssets)
  const numerator = assets.mul(virtualShares);
  return numerator.add(virtualAssets).sub(ONE).div(virtualAssets);
}

/** Convert shares to assets with ceiling rounding (for mint). */
export function convertToAssetsCeil(
  shares: BN,
  totalAssets: BN,
  totalShares: BN,
  decimalsOffset: number
): BN {
  const offset = calculateOffset(decimalsOffset);
  const virtualShares = totalShares.add(offset);
  const virtualAssets = totalAssets.add(ONE);

  // ceil(shares * virtualAssets / virtualShares)
  const numerator = shares.mul(virtualAssets);
  return numerator.add(virtualShares).sub(ONE).div(virtualShares);
}

/** Calculate share price as a floating point number (for display only). */
export function sharePrice(totalAssets: BN, totalShares: BN): number {
  if (totalShares.isZero()) return 1.0;
  return totalAssets.toNumber() / totalShares.toNumber();
}
